### PR TITLE
Lint to remove redundant `clone()`s

### PR DESCRIPTION
**Pull request by sinkuu**
This PR adds lint `redundant_clone`. It suggests to remove redundant `clone()` that clones a owned value that will be dropped without any usage after that.

Real-world example: https://github.com/rust-lang/rust/compare/7b0735a..sinkuu:redundant_clone2

### CODE REVIEW

**Comment by flip1995** [comment]
cc #17

**Comment by sinkuu** [comment]
`cont_eval` errors are apeearing in `indexing_slicing.rs`. This lint causes MIR to be optimized if it has't (since `mir_validated` is stolen sometimes, it uses `optimized_mir` instead), which may trigger const-evals. It should slowdown linting too.

How can I know if `mir_validated` is available? Checking for `no_codegen || !should_codegen()` didn't work...

**Comment by sinkuu** [comment]
Ah, working on optimized MIR was much faster -- using `mir_validated` increased the slowdown to 50%, when it works. Perhaps because less number of BBs we need to visit? I'll just update `indexing_slicing.stderr`.

File: clippy_lints/src/redundant_clone.rs [hunk 1]
```diff
@@ -135,43 +119,23 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for RedundantClone {
...
+            // block.
+            let cloned = unwrap_or_continue!(find_stmt_assigns_to(arg, from_borrow, bbdata.statements.iter().rev()));
 
             // _1 in MIR `{ _2 = &_1; _3 = deref(move _2); } -> { _4 = _3; to_path_buf(move _4); }`
             let referent = if from_deref {
                 let ps = mir.predecessors_for(bb);
+                if ps.len() != 1 {
+                    continue;
+                }
+                let pred_terminator = unwrap_or_continue!(&mir[ps[0]].terminator);
```
Comment (oli-obk) [inline clippy_lints/src/redundant_clone.rs]: ah, this should never be `None` in MIR you get from the compiler. use the `terminator` method to get the value directly.

**Comment by Manishearth** [comment]
Ooh, I've wanted this lint forever! Thanks for writing this!

You may be interested in the set of escape analysis based lints that help
avoid clones, boxes, and vecs -- we have boxed_local that kind of does this
but ideally we'd have a set of lints that handle similar things. There's a
lot of code that trips up on things like this.

**Comment by sinkuu** [comment]
Result of `hyperfine -p "touch path/to/lib.rs" -w 3 -i`:

Package | master [s] | redundant_clone [s] | Slowdown [%]
-- | -- | -- | --
diesel | 25.632 | 26.161 | 2.06
cargo | 13.215 | 14.441 | 9.28
regex | 1.983 | 2.176 | 9.73
serde-json | 0.8877 | 0.9575 | 7.86

File: clippy_lints/src/redundant_clone.rs [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+            if !used_later {
+                let span = terminator.source_info.span;
+                let node = if let mir::ClearCrossCrate::Set(scope_local_data) = &mir.source_scope_local_data {
+                    scope_local_data[terminator.source_info.scope].lint_root
+                } else {
+                    unreachable!()
+                };
+
+                if_chain! {
+                    if !in_macro(span);
```
Comment (oli-obk) [inline clippy_lints/src/redundant_clone.rs]: Move this check all the way to the start of the loop and continue if we are in a macro.

**Comment by oli-obk** [comment]
bors ping

**Comment by mati865** [comment]
Maybe bors is mad at you?
I'm wondering if it will respond to unprivileged person.

bors ping

**Comment by flip1995** [comment]
it was down, now it is back up again 🎉 

bors r=oli-obk

**Comment by notriddle** [comment]
My bad.

... wait, I thought all the official Rust stuff used bors-voyager. What happened?

**Comment by Manishearth** [comment]
That's just crates.io, the official rust stuff uses homu. I think we set up bors-ng because setting up homu requires some infra edits.

**Comment by flip1995** [comment]
bors retry

**Comment by matthiaskrgr** [comment]
bors retry
Just needs a retry I think (not sure if I am privileged)

**Comment by flip1995** [comment]
bors retry

**Comment by flip1995** [comment]
bors ping

**Comment by flip1995** [comment]
It seems that if `retry` was written before another `retry` won't trigger bors again

bors r=oli-obk

**Pull request merged**